File: sdk/examples/typescript/heartbeat_metrics.ts
// Minimal example showing heartbeat (liveness-only) and separate metrics emission
import type {
  AgentIdentity,
  AuthContext,
  EventMessage,
  Heartbeat,
  MetricBatch,
  PapEnvelope,
} from "../typescript/src/index";

function auth(): AuthContext {
  return {
    signature: new Uint8Array([1, 2, 3]),
    algorithm: "ed25519",
    payloadHash: new Uint8Array([4, 5, 6]),
    nonce: crypto.randomUUID(),
    issuedAt: new Date(),
  };
}

const sender: AgentIdentity = { agent: "focus", cluster: "cluster-a" };

export function heartbeatEnvelope(): PapEnvelope<EventMessage> {
  const hb: Heartbeat = {
    cpuPercent: 2.1,
    memoryMb: 128.5,
    uptimeSeconds: 90,
    activeJobs: 0,
  };
  const event: EventMessage = { kind: "event", eventType: "heartbeat", detail: hb };
  return {
    messageId: crypto.randomUUID(),
    sentAt: new Date(),
    sender,
    auth: auth(),
    body: event,
    traceId: "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
    spanId: "abcdef12abcdef12",
  };
}

export function metricsEnvelope(): PapEnvelope<EventMessage> {
  const metrics: MetricBatch = {
    points: [
      { name: "jobs.completed", value: 1, observedAt: new Date(), attributes: { agent: "focus" } },
      { name: "latency.ms", value: 23.7, observedAt: new Date(), attributes: { route: "analyze" } },
    ],
  };
  const event: EventMessage = { kind: "event", eventType: "metric", detail: metrics };
  return {
    messageId: crypto.randomUUID(),
    sentAt: new Date(),
    sender,
    auth: auth(),
    body: event,
    traceId: "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
    spanId: "1234567812345678",
  };
}

// Demo output when run under a TS runtime supporting `crypto.randomUUID()`
console.log("Heartbeat:", heartbeatEnvelope());
console.log("Metrics:", metricsEnvelope());

File: sdk/examples/python/heartbeat_metrics.py
from datetime import datetime
from uuid import uuid4

from pap.envelope import (
    AgentIdentity,
    AuthContext,
    Envelope,
    Event,
    EventType,
    Heartbeat,
    MetricBatch,
    MetricPoint,
)


def make_auth() -> AuthContext:
    # Demo-only placeholder values
    return AuthContext(
        signature=b"demo",
        algorithm="ed25519",
        payload_hash=b"demo",
        nonce=str(uuid4()),
        issued_at=datetime.utcnow(),
    )


def heartbeat_envelope() -> Envelope:
    hb = Heartbeat(cpu_percent=3.2, memory_mb=256.4, uptime_seconds=120.0, active_jobs=0)
    evt = Event(event_type=EventType.HEARTBEAT, heartbeat=hb)
    env = Envelope(
        message_id=str(uuid4()),
        sent_at=datetime.utcnow(),
        sender=AgentIdentity(agent="focus", cluster="cluster-a"),
        auth=make_auth(),
        body=evt,
        # Attach tracing IDs for correlation
        trace_id="0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
        span_id="abcdef12abcdef12",
    )
    return env


def metrics_envelope() -> Envelope:
    points = [
        MetricPoint(name="jobs.completed", value=1.0, observed_at=datetime.utcnow(), attributes={"agent": "focus"}),
        MetricPoint(name="latency.ms", value=25.4, observed_at=datetime.utcnow(), attributes={"route": "analyze"}),
    ]
    batch = MetricBatch(points=points)
    evt = Event(event_type=EventType.METRIC, metrics=batch)
    env = Envelope(
        message_id=str(uuid4()),
        sent_at=datetime.utcnow(),
        sender=AgentIdentity(agent="focus", cluster="cluster-a"),
        auth=make_auth(),
        body=evt,
        trace_id="0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
        span_id="1234567812345678",
    )
    return env


if __name__ == "__main__":
    hb = heartbeat_envelope()
    mx = metrics_envelope()
    # In a real client, these would be serialized and sent to the Proxy
    print("Heartbeat envelope:", hb)
    print("Metrics envelope:", mx)

